src/filter.rs

```diff
@@ -105,152 +105,90 @@ pub fn separable_filter<I, K>(image: &I, h_kernel: &[K], v_kernel: &[K])
 pub fn separable_filter_equal<I, K>(image: &I, kernel: &[K])
         -> VecBuffer<I::Pixel>
     where I: GenericImage,
-          I::Pixel: HasBlack + 'static,
+          I::Pixel: 'static,
           <I::Pixel as Pixel>::Subpixel: ValueInto<K> + Clamp<K>,
           K: Num + Copy {
     separable_filter(image, kernel, kernel)
 }
 
-/// Returns 2d correlation of an image with a 3x3 kernel. Intermediate calculations are
-/// performed at type K, and the results clamped to subpixel type S. Pads by continuity.
-// TODO: factor out the accumulation code from this, horizontal_filter and vertical_filter
-pub fn filter3x3<I, P, K, S>(image: &I, kernel: &[K]) -> VecBuffer<ChannelMap<P, S>>
-    where I: GenericImage<Pixel=P>,
-          P::Subpixel: ValueInto<K>,
-          S: Clamp<K> + Primitive + 'static,
-          P: WithChannel<S> + 'static,
+/// Returns 2d correlation of an image with a kernel. Intermediate calculations are
+/// performed at type K, and the results converted to pixel Q via f. Pads by continuity.
+pub fn filter_kernel<I, K, F, Q>(image: &I, kernel: &[K], 
+                                 k_width: u32, k_height: u32, mut f: F) 
+        -> VecBuffer<Q>
+    where I: GenericImage,
+          <I::Pixel as Pixel>::Subpixel: ValueInto<K>,
+          Q: Pixel + 'static,
+          F: FnMut(&mut Q::Subpixel, K) -> (),
           K: Num + Copy {
 
-    let (width, height) = image.dimensions();
-    let mut out = ImageBuffer::<ChannelMap<P, S>, Vec<S>>::new(width, height);
-    let num_channels = I::Pixel::channel_count() as usize;
+    assert!(kernel.len() as u32 >= k_width * k_height,
+            "incompatible kernel lengths".to_owned());
 
-    // TODO: Should we handle images with height or width < 2? Feels clunky to return Results
-    // TODO: everywhere. Could just document as requirement and leave it at that.
-    for y in 0..height {
-        let y_prev = cmp::max(1, y) - 1;
-        let y_next = cmp::min(height - 2, y) + 1;
-
-        for x in 0..width {
-            let x_prev = cmp::max(1, x) - 1;
-            let x_next = cmp::min(width - 2, x) + 1;
-
-            let mut acc = vec![Zero::zero(); num_channels];
-            let pixels = unsafe {
-                [&image.unsafe_get_pixel(x_prev, y_prev),
-                 &image.unsafe_get_pixel(x     , y_prev),
-                 &image.unsafe_get_pixel(x_next, y_prev),
-                 &image.unsafe_get_pixel(x_prev, y     ),
-                 &image.unsafe_get_pixel(x     , y     ),
-                 &image.unsafe_get_pixel(x_next, y     ),
-                 &image.unsafe_get_pixel(x_prev, y_next),
-                 &image.unsafe_get_pixel(x     , y_next),
-                 &image.unsafe_get_pixel(x_next, y_next),]
-            };
-            for i in 0..9 {
-                accumulate(&mut acc, pixels[i], kernel[i]);
+    let (w, h) = image.dimensions();
+    let mut out = VecBuffer::<Q>::new(w, h);
```

Comment:
This might sound pedantic, but I think we should produce a style guide on names for standard sorts of things. e.g. num_things vs nb_things vs no_things, w vs width, h vs height, etc., etc.

Obviously any choice is fairly arbitrary, and I don't think we need to enforce this religiously, but I always find it easier/more pleasant to read codebases that are consistent. I'm not suggesting that you need to make any changes to this pull request for this reason, but I'll probably do a sweep through the codebase renaming things when we agree on what the conventions should be.